File: verify/claims.rb
# frozen_string_literal: true

require "csv"

root = ARGV[0] || "."
readme = File.read(File.join(root, "README.md"), encoding: "UTF-8")
tsv = CSV.read(File.join(root, "verify", "claims.tsv"), col_sep: "\t", headers: true)

bad = 0
tsv.each do |r|
  a = r["a"].to_f
  b_val = r["b"] == "-" ? nil : r["b"].to_f
  pub = r["published"].to_f
  rule = r["rule"]

  got = case rule
        when "copy"   then a
        when "round0" then a.round(0).to_f
        when "round1" then a.round(1)
        when "round2" then a.round(2)
        when "ratio0" then (a / b_val).round(0).to_f
        when "ratio1" then (a / b_val).round(1)
        else
          puts "  FAIL: unknown rule #{rule}"
          bad += 1
          next
        end

  if (got - pub).abs > 1e-9
    puts "  FAIL: #{r['id']}: published #{pub}, recomputed #{got}"
    bad += 1
  end

  unless readme.include?(r["phrase"])
    puts "  FAIL: #{r['id']}: phrase not in README"
    bad += 1
  end
end

if bad > 0
  puts "Ruby: #{bad} problem(s)"
  exit 1
end
puts "Ruby: #{tsv.size} values reproduced, all phrases found in README"

File: verify/claims.py
import csv, sys, os

root = sys.argv[1] if len(sys.argv) > 1 else "."
readme = open(os.path.join(root, "README.md"), encoding="utf-8").read()

bad = 0
rows = 0
with open(os.path.join(root, "verify", "claims.tsv"), encoding="utf-8") as f:
    for r in csv.DictReader(f, delimiter="\t"):
        rows += 1
        a = float(r["a"])
        b = float(r["b"]) if r["b"] != "-" else None
        pub = float(r["published"])
        rule = r["rule"]

        if rule == "copy":
            got = a
        elif rule == "round0":
            got = round(a)
        elif rule == "round1":
            got = round(a, 1)
        elif rule == "round2":
            got = round(a, 2)
        elif rule == "ratio0":
            got = round(a / b)
        elif rule == "ratio1":
            got = round(a / b, 1)
        else:
            print(f"  FAIL: unknown rule {rule}")
            bad += 1
            continue

        if abs(got - pub) > 1e-9:
            print(f"  FAIL: {r['id']}: published {pub}, recomputed {got}")
            bad += 1

        if r["phrase"] not in readme:
            print(f"  FAIL: {r['id']}: phrase not in README")
            bad += 1

if bad:
    print(f"Python: {bad} problem(s)")
    sys.exit(1)
print(f"Python: {rows} values reproduced, all phrases found in README")
